pallets/drop3/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+		/// A reward pool with \[ id \] was started, either by admin or owner
+		RewardPoolStarted { id: T::PoolId },
+		/// A reward pool with \[ id \] was stopped, either by admin or owner
+		RewardPoolStopped { id: T::PoolId },
+		/// A reward pool with \[ id, name, owner \] was removed, either by admin or owner
+		RewardPoolRemoved { id: T::PoolId, name: Vec<u8>, owner: T::AccountId },
+		/// A reward pool with \[ id, name, owner \] was proposed
+		RewardPoolProposed { id: T::PoolId, name: Vec<u8>, owner: T::AccountId },
+		/// An \[ amount \] of reward was sent to \[ to \]
+		RewardSent { to: T::AccountId, amount: T::Balance },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Error when the caller account is not the admin
+		RequireAdmin,
+		/// Error when the caller account is not the reward pool owner
+		RequireRewardPoolOwner,
+		/// Error when the caller account is not the reward pool owner or admin
+		RequireAdminOrRewardPoolOwner,
+		/// Error when a reward pool can't be found
+		NoSuchRewardPool,
+		/// Error when the sender doesn't have enough reserved balance
+		InsufficientReservedBalance,
+		/// Error when `total` amount is 0 when proposing reward pool
+		InvalidTotalBalance,
+		/// Error when the remaning of a reward pool is not enough
+		InsufficientRemain,
+		/// Error when start_at < end_at when proposing reward pool
+		InvalidProposedBlock,
+		/// Error when the reward pool is unapproved
+		RewardPoolUnapproved,
+		/// Error when the reward pool is first approved then rejected
+		RewardPoolAlreadyApproved,
+		/// Error when the reward pool is stopped
+		RewardPoolStopped,
+		/// Error when the reward pool is runing before `start_at`
+		RewardPoolRanTooEarly,
+		/// Error when the reward pool is runing after `end_at`
+		RewardPoolRanTooLate,
+		/// Error of unexpected unmoved amount when calling repatriate_reserved
+		UnexpectedUnMovedAmount,
+		/// Error when no vacant PoolId can be acquired
+		NoVacantPoolId,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Change the admin account
+		/// similar to sudo.set_key, the old account will be supplied in event
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_admin())]
+		pub fn set_admin(origin: OriginFor<T>, new: T::AccountId) -> DispatchResultWithPostInfo {
+			T::SetAdminOrigin::ensure_origin(origin)?;
+			Self::deposit_event(Event::AdminChanged { old_admin: Self::admin() });
+			<Admin<T>>::put(new);
+			// Do not pay a fee
+			Ok(Pays::No.into())
+		}
+
+		/// Approve a RewardPool proposal, must be called from admin
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::approve_reward_pool())]
+		pub fn approve_reward_pool(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(sender == Self::admin().unwrap(), Error::<T>::RequireAdmin);
+			ensure!(RewardPools::<T>::contains_key(id), Error::<T>::NoSuchRewardPool);
+
+			RewardPools::<T>::mutate(id, |pool| {
```

Comment:
check if already approved before

Reply:
> check if already approved before

Maybe I'll keep this.
From code logic, multiple calls of `approve_reward_pool()` should be OK and I just leave the event as it is.
I could have added an if-check inside `mutate`, but the extra DB read in all cases is probably not worth it:
`100% 1 * write` vs `50% 1* read + 50% 1 * (read + write)`